NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,729 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+	double& Get( int threadIndex ) { return localInertia[threadIndex]; }
+	double GetSum() const;
+
+private:
+	CArray<double> localInertia{};
+};
+
+double CInertia::GetSum() const
+{
+	double inertia = 0;
+	for( int i = 0; i < localInertia.Size(); ++i ) {
+		inertia += localInertia[i];
+	}
+	return inertia;
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+// Task which processes a set of elements on a single thread
+struct CKMeansClustering::IThreadTask {
+public:
+	virtual ~IThreadTask() {}
+	virtual int Size() const { return VectorSize; }
+
+	int ThreadCount() const { return Owner.params.ThreadCount; }
+	void CallRun( int threadIndex, int align = 1 );
+protected:
+	IThreadTask( const CKMeansClustering& owner, int size ) :
+		VectorSize( size ),
+		Owner( owner )
+	{}
+
+	virtual void Run( int threadIndex, int index, int count ) = 0;
+
+	const int VectorSize;
+	const CKMeansClustering& Owner;
+};
+
+void CKMeansClustering::IThreadTask::CallRun( int threadIndex, int align )
+{
+	int index = 0;
+	int count = 0;
+	if( GetTaskIndexAndCount( ThreadCount(), threadIndex, Size(), align, index, count ) ) {
+		Run( threadIndex, index, count );
+	}
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+// Task which processes a set of elements one-by-one on a single thread
+struct CKMeansClustering::IThreadSubTask : public CKMeansClustering::IThreadTask {
+public:
+	int Size() const override final { return Matrix ? Matrix->Height : VectorSize; }
+
+	const CFloatMatrixDesc* const Matrix;
+protected:
+	IThreadSubTask( const CKMeansClustering& owner, int size ) :
+		IThreadTask( owner, size ),
+		Matrix( nullptr )
+	{}
+	IThreadSubTask( const CKMeansClustering& owner, const CFloatMatrixDesc* matrix ) :
+		IThreadTask( owner, 0 ),
+		Matrix( matrix )
+	{}
+
+	void Run( int threadIndex, int index, int count ) override final;
+	virtual void RunOnElement( int threadIndex, int index ) = 0;
+};
+
+void CKMeansClustering::IThreadSubTask::Run( int threadIndex, int index, int count )
+{
+	const int after_last = index + count;
+	for( int i = index; i < after_last; ++i ) {
+		RunOnElement( threadIndex, i );
+	}
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CKMeansClustering::CClassifyAllThreadTask final : public CKMeansClustering::IThreadSubTask {
+	CClassifyAllThreadTask( const CKMeansClustering& owner, const CFloatMatrixDesc& matrix ) :
+		IThreadSubTask( owner, &matrix ),
+		Inertia( ThreadCount() )
+	{ DataCluster.SetBufferSize( Size() ); }
+
+	CArray<int> DataCluster{}; // the cluster for this element
+	CInertia Inertia;
+protected:
+	void RunOnElement( int threadIndex, int index ) override
+	{ DataCluster[index] = Owner.findNearestCluster( *Matrix, index, Inertia.Get( threadIndex ) ); }
```

Comment:
In that case it's better to move `findNearestCluster` from owner to here. (The only exception if it heavily relies on other fields/methods of kmeans).

Reply:
fixed